File: examples/fastapi-app/app/services/microservices.py
"""Microservices user registration business logic.

Pure functions that create user accounts, set up billing profiles,
initialize preferences, send welcome sequences, and update user status.
No Tasker types -- just plain dicts in, typed models out.

Diamond dependency pattern:
  CreateUserAccount
       |-- SetupBillingProfile  --|
       |-- InitializePreferences -|
                                   |-- SendWelcomeSequence -> UpdateUserStatus
"""

from __future__ import annotations

import hashlib
import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from tasker_core.errors import PermanentError, RetryableError

from .types import (
    CreateUserAccountInput,
    MicroservicesCreateUserResult,
    MicroservicesInitPreferencesResult,
    MicroservicesSendWelcomeResult,
    MicroservicesSetupBillingResult,
    MicroservicesUpdateStatusResult,
)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

RESERVED_USERNAMES = {"admin", "root", "system", "support", "test"}

PLAN_PRICING = {
    "starter": {"monthly_price": 0.0, "annual_price": 0.0, "trial_days": 0},
    "professional": {"monthly_price": 29.99, "annual_price": 299.99, "trial_days": 14},
    "enterprise": {"monthly_price": 99.99, "annual_price": 999.99, "trial_days": 30},
}

PLAN_LIMITS = {
    "starter": {"api_calls": 1000, "storage_gb": 1, "team_members": 1},
    "professional": {"api_calls": 50000, "storage_gb": 50, "team_members": 10},
    "enterprise": {"api_calls": -1, "storage_gb": 500, "team_members": -1},
}


# ---------------------------------------------------------------------------
# Service functions
# ---------------------------------------------------------------------------


def create_user_account(
    input: CreateUserAccountInput,
) -> MicroservicesCreateUserResult:
    """Validate user input, generate a unique user ID, and create the initial account."""
    email = input.email
    username = input.username
    plan = input.plan or "starter"
    referral_code = input.referral_code

    # Fields guaranteed non-None by @model_validator (raises PermanentError if missing)
    assert email is not None
    assert username is not None

    if "@" not in email:
        raise PermanentError(f"Invalid email address: {email}")

    full_name = username
    if len(full_name.strip()) < 2:
        raise PermanentError("Name must be at least 2 characters")

    if username.lower() in RESERVED_USERNAMES:
        raise PermanentError(f"Username '{username}' is reserved")

    internal_id = f"usr_{uuid.uuid4().hex[:12]}"
    derived_username = email.split("@")[0].lower()
    user_id = username or derived_username
    verification_token = hashlib.sha256(
        f"{internal_id}:{email}:{datetime.now(timezone.utc).isoformat()}".encode()
    ).hexdigest()[:32]

    return MicroservicesCreateUserResult(
        user_id=user_id,
        email=email,
        name=full_name,
        plan=plan,
        phone=None,
        source="web",
        status="created",
        internal_id=internal_id,
        username=derived_username,
        full_name=full_name,
        referral_code=referral_code,
        verification_token=verification_token,
        email_verified=False,
        account_status="pending_verification",
        created_at=datetime.now(timezone.utc).isoformat(),
    )


def setup_billing_profile(
    user_data: MicroservicesCreateUserResult,
) -> MicroservicesSetupBillingResult:
    """Configure plan-specific billing tiers with pricing and features."""
    if user_data is None:
        raise PermanentError("Missing create_user_account dependency")

    user_id = user_data.user_id
    plan = user_data.plan or "starter"
    internal_id = user_data.internal_id

    pricing = PLAN_PRICING.get(plan, PLAN_PRICING["starter"])
    limits = PLAN_LIMITS.get(plan, PLAN_LIMITS["starter"])
    billing_required = pricing["monthly_price"] > 0

    billing_id = f"bill_{uuid.uuid4().hex[:12]}"
    subscription_id = f"sub_{uuid.uuid4().hex[:12]}"

    trial_end = None
    next_billing_date = None
    if pricing["trial_days"] > 0:
        trial_end = (
            datetime.now(timezone.utc) + timedelta(days=pricing["trial_days"])
        ).isoformat()

    if billing_required:
        next_billing_date = (
            datetime.now(timezone.utc) + timedelta(days=30)
        ).isoformat()

    now = datetime.now(timezone.utc).isoformat()

    return MicroservicesSetupBillingResult(
        billing_id=billing_id,
        user_id=user_id,
        plan=plan,
        price=pricing["monthly_price"],
        currency="USD",
        billing_cycle="monthly",
        features=list(limits.keys()),
        status="active" if billing_required else "skipped_free_plan",
        billing_required=billing_required,
        next_billing_date=next_billing_date,
        subscription_id=subscription_id,
        user_internal_id=internal_id,
        pricing=pricing,
        limits=limits,
        billing_status="trial" if trial_end else "active",
        trial_end=trial_end,
        payment_method_required=billing_required,
        created_at=now,
    )


def initialize_preferences(
    user_data: MicroservicesCreateUserResult,
    custom_prefs: dict[str, Any] | None,
) -> MicroservicesInitPreferencesResult:
    """Initialize user preferences with plan-appropriate defaults."""
    if user_data is None:
        raise PermanentError("Missing create_user_account dependency")

    user_id = user_data.user_id
    plan = user_data.plan or "starter"
    internal_id = user_data.internal_id
    custom_prefs = custom_prefs or {}

    notifications = {
        "email_updates": True,
        "marketing_emails": plan != "enterprise",
        "weekly_digest": True,
        "security_alerts": True,
        "product_updates": True,
        "billing_alerts": plan != "starter",
    }

    ui_settings = {
        "theme": "light",
        "language": "en",
        "timezone": "UTC",
        "date_format": "YYYY-MM-DD",
        "items_per_page": 25,
        "sidebar_collapsed": False,
    }

    feature_flags = {
        "beta_features": plan == "enterprise",
        "advanced_analytics": plan in ("professional", "enterprise"),
        "api_access": plan in ("professional", "enterprise"),
        "custom_integrations": plan == "enterprise",
        "priority_support": plan == "enterprise",
        "export_data": True,
    }

    default_prefs = {**notifications, **ui_settings}
    preferences = {**default_prefs, **custom_prefs}

    preferences_id = f"pref_{uuid.uuid4().hex[:12]}"
    now = datetime.now(timezone.utc).isoformat()

    return MicroservicesInitPreferencesResult(
        preferences_id=preferences_id,
        user_id=user_id,
        plan=plan,
        preferences=preferences,
        defaults_applied=len(default_prefs),
        customizations=len(custom_prefs),
        status="active",
        user_internal_id=internal_id,
        notifications=notifications,
        ui_settings=ui_settings,
        feature_flags=feature_flags,
        onboarding_completed=False,
        created_at=now,
        updated_at=now,
    )


def send_welcome_sequence(
    user_data: MicroservicesCreateUserResult,
    billing_data: MicroservicesSetupBillingResult,
    prefs_data: MicroservicesInitPreferencesResult,
) -> MicroservicesSendWelcomeResult:
    """Send a multi-channel welcome sequence to the new user."""
    if not all([user_data, billing_data, prefs_data]):
        raise PermanentError("Missing upstream dependency results")

    user_id = user_data.user_id
    email = user_data.email
    plan = user_data.plan or "starter"
    full_name = user_data.full_name or user_data.name
    verification_token = user_data.verification_token
    trial_end = billing_data.trial_end
    prefs = prefs_data.preferences or {}

    messages_sent_list: list[dict[str, Any]] = []

    # Welcome email (if email notifications enabled)
    if prefs.get("email_updates", True) or prefs.get("email_notifications", True):
        welcome_msg_id = f"msg_{uuid.uuid4().hex[:16]}"
        messages_sent_list.append(
            {
                "message_id": welcome_msg_id,
                "channel": "email",
                "recipient": email,
                "template": "welcome_email_v3",
                "subject": f"Welcome to Tasker, {full_name}!",
                "status": "sent",
            }
        )

    # Verification email
    verify_msg_id = f"msg_{uuid.uuid4().hex[:16]}"
    messages_sent_list.append(
        {
            "message_id": verify_msg_id,
            "channel": "email",
            "recipient": email,
            "template": "email_verification",
            "subject": "Verify your email address",
            "verification_link": f"https://app.example.com/verify/{verification_token}",
            "status": "sent",
        }
    )

    # In-app onboarding
    onboard_msg_id = f"msg_{uuid.uuid4().hex[:16]}"
    messages_sent_list.append(
        {
            "message_id": onboard_msg_id,
            "channel": "in_app",
            "template": "onboarding_tour",
            "status": "queued",
        }
    )

    # Trial notification if applicable
    if trial_end:
        trial_msg_id = f"msg_{uuid.uuid4().hex[:16]}"
        messages_sent_list.append(
            {
                "message_id": trial_msg_id,
                "channel": "email",
                "recipient": email,
                "template": "trial_started",
                "subject": f"Your {plan} trial has started",
                "trial_end": trial_end,
                "status": "sent",
            }
        )

    channels_used = list({m["channel"] for m in messages_sent_list})
    welcome_sequence_id = f"welcome_{uuid.uuid4().hex[:12]}"
    now = datetime.now(timezone.utc).isoformat()

    return MicroservicesSendWelcomeResult(
        user_id=user_id,
        plan=plan,
        channels_used=channels_used,
        messages_sent=len(messages_sent_list),
        welcome_sequence_id=welcome_sequence_id,
        status="sent",
        messages_sent_details=messages_sent_list,
        total_messages=len(messages_sent_list),
        sequence_id=f"seq_{uuid.uuid4().hex[:12]}",
        sent_at=now,
    )


def update_user_status(
    user_data: MicroservicesCreateUserResult,
    billing_data: MicroservicesSetupBillingResult,
    preferences_data: MicroservicesInitPreferencesResult,
    welcome_data: MicroservicesSendWelcomeResult,
) -> MicroservicesUpdateStatusResult:
    """Finalize user registration by activating the account."""
    if not all([user_data, billing_data, preferences_data, welcome_data]):
        raise PermanentError("Missing upstream dependency results")

    user_id = user_data.user_id
    plan = user_data.plan or "starter"
    email = user_data.email
    internal_id = user_data.internal_id
    billing_id = billing_data.billing_id
    subscription_id = billing_data.subscription_id
    messages_sent = welcome_data.total_messages if welcome_data.total_messages is not None else (welcome_data.messages_sent or 0)

    registration_summary: dict[str, Any] = {
        "user_id": user_id,
        "email": email,
        "plan": plan,
        "registration_status": "complete",
    }
    if plan != "starter" and billing_data.billing_id:
        registration_summary["billing_id"] = billing_data.billing_id
        registration_summary["next_billing_date"] = billing_data.next_billing_date
    prefs = preferences_data.preferences or {}
    registration_summary["preferences_count"] = len(prefs) if isinstance(prefs, dict) else 0
    registration_summary["welcome_sent"] = True
    registration_summary["notification_channels"] = welcome_data.channels_used or []
    registration_summary["user_created_at"] = user_data.created_at
    registration_summary["registration_completed_at"] = datetime.now(timezone.utc).isoformat()

    now = datetime.now(timezone.utc).isoformat()

    return MicroservicesUpdateStatusResult(
        user_id=user_id,
        status="active",
        plan=plan,
        registration_summary=registration_summary,
        activation_timestamp=now,
        all_services_coordinated=True,
        services_completed=[
            "user_service",
            "billing_service",
            "preferences_service",
            "notification_service",
        ],
        internal_id=internal_id,
        email=email,
        account_status="active",
        billing_id=billing_id,
        subscription_id=subscription_id,
        onboarding_status="in_progress",
        welcome_messages_sent=messages_sent,
        registration_complete=True,
        activated_at=now,
    )

File: examples/bun-app/src/services/microservices.ts
/**
 * Microservices user registration business logic.
 *
 * Pure functions that create user accounts, set up billing profiles,
 * initialize preferences, send welcome sequences, and update user status.
 * No Tasker types -- just plain objects in, plain objects out.
 *
 * Diamond dependency pattern:
 *   CreateUserAccount
 *        |-- SetupBillingProfile  --|
 *        |-- InitializePreferences -|
 *                                    |-- SendWelcomeSequence -> UpdateUserStatus
 */

import { PermanentError } from '@tasker-systems/tasker';

import type {
  CreateUserAccountInput,
  BillingTier,
  MicroservicesCreateUserResult,
  MicroservicesSetupBillingResult,
  MicroservicesInitPreferencesResult,
  MicroservicesSendWelcomeResult,
  MicroservicesUpdateStatusResult,
} from './types';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const BILLING_TIERS: Record<string, BillingTier> = {
  free: { price: 0, features: ['basic_features'], billing_required: false },
  pro: {
    price: 29.99,
    features: ['basic_features', 'advanced_analytics'],
    billing_required: true,
  },
  enterprise: {
    price: 299.99,
    features: [
      'basic_features',
      'advanced_analytics',
      'priority_support',
      'custom_integrations',
    ],
    billing_required: true,
  },
  basic: { price: 9.99, features: ['basic_features'], billing_required: true },
  standard: {
    price: 29.99,
    features: ['basic_features', 'advanced_analytics'],
    billing_required: true,
  },
  premium: {
    price: 99.99,
    features: ['basic_features', 'advanced_analytics', 'priority_support'],
    billing_required: true,
  },
};

const DEFAULT_PREFERENCES: Record<string, Record<string, unknown>> = {
  free: {
    email_notifications: true,
    marketing_emails: false,
    product_updates: true,
    weekly_digest: false,
    theme: 'light',
    language: 'en',
    timezone: 'UTC',
  },
  pro: {
    email_notifications: true,
    marketing_emails: true,
    product_updates: true,
    weekly_digest: true,
    theme: 'dark',
    language: 'en',
    timezone: 'UTC',
    api_notifications: true,
  },
  enterprise: {
    email_notifications: true,
    marketing_emails: true,
    product_updates: true,
    weekly_digest: true,
    theme: 'dark',
    language: 'en',
    timezone: 'UTC',
    api_notifications: true,
    audit_logs: true,
    advanced_reports: true,
  },
};

// ---------------------------------------------------------------------------
// Service functions
// ---------------------------------------------------------------------------

export function createUserAccount(
  input: CreateUserAccountInput,
): MicroservicesCreateUserResult {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(input.email)) {
    throw new PermanentError(`Invalid email format: ${input.email}`);
  }

  const source = (input.metadata?.referral_source as string) || 'web';
  const userId = crypto.randomUUID();
  const now = new Date().toISOString();
  const apiKey = `ak_${crypto.randomUUID().replace(/-/g, '').substring(0, 32)}`;

  return {
    user_id: userId,
    email: input.email,
    name: input.username,
    plan: input.plan || 'free',
    phone: null,
    source,
    status: 'created',
    created_at: now,
    api_key: apiKey,
    auth_provider: 'internal',
  };
}

export function setupBilling(userData: Record<string, unknown>): MicroservicesSetupBillingResult {
  if (!userData) {
    throw new PermanentError('User data not found from create_user_account step');
  }

  const userId = userData.user_id as string;
  const plan = (userData.plan as string) || 'free';
  const tierConfig = BILLING_TIERS[plan] || BILLING_TIERS.free;

  if (tierConfig.billing_required) {
    const now = new Date();
    const nextBilling = new Date(now.getTime() + 30 * 24 * 60 * 60 * 1000);

    return {
      billing_id: `billing_${crypto.randomUUID().replace(/-/g, '').substring(0, 12)}`,
      user_id: userId,
      plan,
      price: tierConfig.price,
      currency: 'USD',
      billing_cycle: 'monthly',
      features: tierConfig.features,
      status: 'active',
      next_billing_date: nextBilling.toISOString(),
      created_at: now.toISOString(),
    };
  } else {
    return {
      user_id: userId,
      plan,
      billing_required: false,
      status: 'skipped_free_plan',
      message: 'Free plan users do not require billing setup',
    };
  }
}

export function initPreferences(
  userData: Record<string, unknown>,
  metadata: Record<string, unknown> | undefined,
): MicroservicesInitPreferencesResult {
  if (!userData) {
    throw new PermanentError('User data not found from create_user_account step');
  }

  const userId = userData.user_id as string;
  const plan = (userData.plan as string) || 'free';
  const customPrefs = ((metadata?.preferences as Record<string, unknown>) || {});
  const defaultPrefs = DEFAULT_PREFERENCES[plan] || DEFAULT_PREFERENCES.free;
  const finalPrefs = { ...defaultPrefs, ...customPrefs };
  const now = new Date().toISOString();

  return {
    preferences_id: crypto.randomUUID(),
    user_id: userId,
    plan,
    preferences: finalPrefs,
    defaults_applied: Object.keys(defaultPrefs).length,
    customizations: Object.keys(customPrefs).length,
    status: 'active',
    created_at: now,
    updated_at: now,
  };
}

export function sendWelcome(
  userData: Record<string, unknown>,
  billingData: Record<string, unknown>,
  preferencesData: Record<string, unknown>,
): MicroservicesSendWelcomeResult {
  const missing: string[] = [];
  if (!userData) missing.push('create_user_account');
  if (!billingData) missing.push('setup_billing_profile');
  if (!preferencesData) missing.push('initialize_preferences');

  if (missing.length > 0) {
    throw new PermanentError(`Missing results from steps: ${missing.join(', ')}`);
  }

  const userId = userData.user_id as string;
  const email = userData.email as string;
  const plan = (userData.plan as string) || 'free';
  const prefs = (preferencesData.preferences || {}) as Record<string, unknown>;

  const channelsUsed: string[] = [];
  const now = new Date().toISOString();

  if (prefs.email_notifications !== false) {
    channelsUsed.push('email');
  }

  channelsUsed.push('in_app');

  if (plan === 'enterprise') {
    channelsUsed.push('sms');
  }

  return {
    user_id: userId,
    plan,
    channels_used: channelsUsed,
    messages_sent: channelsUsed.length,
    welcome_sequence_id: crypto.randomUUID(),
    status: 'sent',
    sent_at: now,
    recipient: email,
  };
}

export function updateStatus(
  userData: Record<string, unknown>,
  billingData: Record<string, unknown>,
  preferencesData: Record<string, unknown>,
  welcomeData: Record<string, unknown>,
): MicroservicesUpdateStatusResult {
  const missing: string[] = [];
  if (!userData) missing.push('create_user_account');
  if (!billingData) missing.push('setup_billing_profile');
  if (!preferencesData) missing.push('initialize_preferences');
  if (!welcomeData) missing.push('send_welcome_sequence');

  if (missing.length > 0) {
    throw new PermanentError(
      `Cannot complete registration: missing results from steps: ${missing.join(', ')}`,
    );
  }

  const userId = userData.user_id as string;
  const email = userData.email as string;
  const plan = (userData.plan as string) || 'free';

  const prefs = (preferencesData.preferences || {}) as Record<string, unknown>;
  const registrationSummary: Record<string, unknown> = {
    user_id: userId,
    email,
    plan,
    registration_status: 'complete',
    preferences_count: Object.keys(prefs).length,
    welcome_sent: true,
    notification_channels: welcomeData.channels_used,
    user_created_at: userData.created_at,
    registration_completed_at: new Date().toISOString(),
  };

  if (plan !== 'free' && billingData.billing_id) {
    registrationSummary.billing_id = billingData.billing_id;
    registrationSummary.next_billing_date = billingData.next_billing_date;
  }

  const now = new Date().toISOString();

  return {
    user_id: userId,
    status: 'active',
    plan,
    registration_summary: registrationSummary,
    activation_timestamp: now,
    all_services_coordinated: true,
    services_completed: [
      'user_service',
      'billing_service',
      'preferences_service',
      'notification_service',
    ],
  };
}
